### PR TITLE
Upgrade leptos-spin to support leptos 0.6

### PR DESCRIPTION
**Pull request by benwis**
This is my draft PR to make leptos-spin compatible with leptos 0.6. Because we've reworked server functions, we now need a proc macro crate, so I've made this a workspace.
Should address #16 

This does not compile yet, I'm stuck on two issues
1. I can't mutate status on a Request
2. I can't figure out how to pass a stream to an OutgoingResponse or parse the body as a stream from IncomingRequest

### CODE REVIEW

**Comment by benwis** [comment]
@itowlson I've finished the changes for leptos 0.6, and believe they're ready for merging!

File: leptos-spin/templates/leptos-ssr/content/Cargo.toml
Comment (itowlson) [inline leptos-spin/templates/leptos-ssr/content/Cargo.toml:1]: I'd imagine this needs to be updated to refer to the Leptos 0.6 crates?

File: leptos-spin/templates/leptos-ssr/content/.gitignore
Comment (itowlson) [inline leptos-spin/templates/leptos-ssr/content/.gitignore:1]: `templates` needs to remain at the top level of the repo - it's not subsidiary to the `leptos-spin` crate.

**Review by itowlson — CHANGES_REQUESTED** [review]
I'm afraid most of this is beyond my knowledge to review properly, and I can only thank you for the deep knowledge and extensive effort you've brought to this.  You mentioned once that it was nice having an integration that you didn't have to write, and now, well, you've had to write it I fear!

There is one niggle and that is the Spin template.  As noted in the comments, this has to stay at the top level; but also its Cargo.toml will need to be updated with Leptos references that match the `leptos-spin` ones.  To test the template, you can run `spin templates install --git https://github.com/benwis/leptos-spin --btanch main --upgrade` followed by `spin new -t leptos-ssr <some-test-project-name>`.  If you don't fancy tackling this, give me a shout and I can pick it up and send you a PR for it.  Thanks!

**Comment by itowlson** [comment]
Looks like there are a couple of conflicts too, hopefully easy ports from the file moves.

File: leptos-spin-macro/src/lib.rs [hunk 1]
```diff
@@ -17,3 +17,4 @@ pub fn server(args: proc_macro::TokenStream, s: TokenStream) -> TokenStream {
         Ok(s) => s.to_token_stream().into(),
     }
 }
+/Users/sam/Projects/start-spin/src
```
Comment (itowlson) [inline leptos-spin-macro/src/lib.rs]: I think this is a rogue paste

File: templates/leptos-ssr/content/Cargo.toml.template
Comment (itowlson) [inline templates/leptos-ssr/content/Cargo.toml.template:1]: This seems to have been renamed.  Does that work?  I don't remember Spin having a feature to remove a `.template` suffix but maybe I'm misremembering.

File: templates/leptos-ssr/content/src/server.rs [hunk 2]
```diff
@@ -1,19 +1,21 @@
 use leptos::ServerFn;
-use spin_sdk::http::{ResponseOutparam, IncomingRequest};
+use leptos_spin::{render_best_match_to_stream, RouteTable, server_fn::register_explicit};
+use spin_sdk::http::{IncomingRequest, ResponseOutparam};
 use spin_sdk::http_component;
-use leptos_spin::{render_best_match_to_stream, RouteTable};
 
 #[http_component]
-async fn handle_{{project-name | snake_case}}(req: IncomingRequest, resp_out: ResponseOutparam) {
+async fn handle_{{crate_name}}(req: IncomingRequest, resp_out: ResponseOutparam) {
```
Comment (itowlson) [inline templates/leptos-ssr/content/src/server.rs]: I think you need to retain the `snake_case` filter here.  Also does `crate_name` work?  `project-name` is provided from the `spin new` command.

File: templates/leptos-ssr/content/src/app.rs
Comment (itowlson) [inline templates/leptos-ssr/content/src/app.rs:1]: Is it intentional that this is deleted?  It used to provide all the starter UI stuff.  Without it I'm not sure what AppRouter in `server.rs` refers to.

**Review by itowlson — COMMENTED** [review]
Not sure if this is an interim commit?  I realised I made your life unnecessarily difficult by advising you to install from Git... it would of course be more efficient to install from your working copy, using `spin templates install --dir . --upgrade` (or the path to the working copy root if you're not in that directory).  I'm very sorry - mind in a total fog today.

File: leptos-spin-macro/src/lib.rs (hunk 1 above)
Comment (benwis) [reply leptos-spin-macro/src/lib.rs]: Yeah, I thought so too. I need to remove that, they didn't get to it

File: templates/leptos-ssr/content/Cargo.toml.template
Comment (benwis) [reply templates/leptos-ssr/content/Cargo.toml.template:1]: That's a good question, this came from a PR and I haven't tried it too much

**Comment by benwis** [comment]
To be perfectly honest I did not test the template chances, they were done by @sjud and I've been working off the leptos template

**Comment by benwis** [comment]
@itowlson My github fu might be bad, but it won't show me what the conflicts are because I don't have write access to the repo

**Comment by itowlson** [comment]
Huh.  It won't show _me_ what the conflicts are either.  "Too complex to resolve in the web editor" apparently.

image

Could you try doing a rebase locally?

Also, let me know if you'd like me to pick up the template chanmges - happy to send a PR.

**Comment by benwis** [comment]
> Huh. It won't show _me_ what the conflicts are either. "Too complex to resolve in the web editor" apparently.
> 
> image
> 
> Could you try doing a rebase locally?
> 
> Also, let me know if you'd like me to pick up the template chanmges - happy to send a PR.

Sure, I'd love you to handle those. I'll try rebasing against the origin repo

File: leptos-spin-macro/src/lib.rs (hunk 1 above)
Comment (sjud) [reply leptos-spin-macro/src/lib.rs]: Sorry, I've been distracted. Yes rogue paste.

**Review by itowlson — APPROVED** [review]

**Pull request merged**